File: Set Intersection Size At Least Two/Set Intersection Size At Least Two.py
# Solution for Set Intersection Size At Least Two in PY

class Solution:
    def intersectionSizeTwo(self, intervals: List[List[int]]) -> int:

        # Sort by end ASC, start DESC if end same
        intervals.sort(key=lambda x: (x[1], -x[0]))

        second_last = intervals[0][1] - 1
        last = intervals[0][1]

        selected_count = 2

        for i in range(1, len(intervals)):
            start, end = intervals[i]

            # Case 1: interval already contains both points
            if second_last >= start:
                continue

            # Case 2: interval contains only `last`
            elif last >= start:
                second_last = last
                last = end
                selected_count += 1

            # Case 3: contains none → add 2 new points
            else:
                second_last = end - 1
                last = end
                selected_count += 2

        return selected_count

File: Set Intersection Size At Least Two/Set Intersection Size At Least Two.java
// Solution for Set Intersection Size At Least Two in JAVA

class Solution {
    public int intersectionSizeTwo(int[][] intervals) {

        // Sort by end ASC, and if end same, start DESC
        Arrays.sort(intervals, (a, b) -> {
            if (a[1] == b[1]) return b[0] - a[0];
            return a[1] - b[1];
        });

        int secondLast = intervals[0][1] - 1;
        int last = intervals[0][1];

        int selectedCount = 2; // already picked 2 points for the first interval

        for (int i = 1; i < intervals.length; i++) {
            int start = intervals[i][0];
            int end = intervals[i][1];

            // Case 1: interval already contains both points
            if (secondLast >= start) {
                continue;
            }
            // Case 2: interval contains only `last`
            else if (last >= start) {
                secondLast = last;
                last = end;
                selectedCount++;
            }
            // Case 3: contains none → add 2 new points
            else {
                secondLast = end - 1;
                last = end;
                selectedCount += 2;
            }
        }
        return selectedCount;
    }
}
